File: backend/app/routers/project_comments.py
"""
API Router for Project Discussions (#930)
"""

from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, Query, Response, status
from sqlalchemy.orm import Session

from app.dependencies import (
    get_current_active_user,
    get_database,
    get_optional_current_user,
)
from app.models.project_comment import ProjectComment
from app.models.user import User
from app.schemas.project_comment import (
    DELETED_BODY_PLACEHOLDER,
    ProjectCommentCreate,
    ProjectCommentList,
    ProjectCommentResponse,
    ProjectCommentThread,
    ProjectCommentUpdate,
)
from app.services.project_comment_service import (
    DEFAULT_PAGE_SIZE,
    MAX_PAGE_SIZE,
    ProjectCommentService,
)

router = APIRouter(
    prefix="/projects/{project_id}/comments",
    tags=["Project Comments"],
)


def _to_response(comment: ProjectComment) -> ProjectCommentResponse:
    """Serialise a comment, tombstoning the body if it was deleted.

    The row is kept so replies do not lose their parent, but the text is not
    handed back -- "deleted" has to mean the words are gone.
    """
    deleted = comment.deleted_at is not None

    return ProjectCommentResponse(
        id=comment.id,
        project_id=comment.project_id,
        parent_id=comment.parent_id,
        body=DELETED_BODY_PLACEHOLDER if deleted else comment.body,
        author=None if deleted else comment.author,
        is_edited=comment.is_edited,
        edited_at=comment.edited_at,
        is_deleted=deleted,
        created_at=comment.created_at,
        updated_at=comment.updated_at,
    )


def _to_thread(comment: ProjectComment) -> ProjectCommentThread:
    replies = ProjectCommentService.visible_replies(comment)
    base = _to_response(comment)

    return ProjectCommentThread(
        **base.model_dump(),
        replies=[_to_response(reply) for reply in replies],
        reply_count=len(replies),
    )


@router.get(
    "",
    response_model=ProjectCommentList,
    summary="List project comments",
    description=(
        "Top-level comments for a project, newest first, each with its "
        "replies. Comments on a private project are only visible to people "
        "who can see the project."
    ),
)
@router.get("/", response_model=ProjectCommentList, include_in_schema=False)
def list_comments(
    project_id: uuid.UUID,
    limit: int = Query(
        DEFAULT_PAGE_SIZE,
        ge=1,
        le=MAX_PAGE_SIZE,
        description="Top-level comments per page.",
    ),
    offset: int = Query(0, ge=0),
    db: Session = Depends(get_database),
    current_user: User | None = Depends(get_optional_current_user),
) -> ProjectCommentList:
    project = ProjectCommentService.get_project_or_404(db, project_id)
    ProjectCommentService.require_can_view(db, project, current_user)

    comments, total = ProjectCommentService.list_thread(
        db, project_id, limit=limit, offset=offset
    )

    return ProjectCommentList(
        items=[_to_thread(comment) for comment in comments],
        total=total,
        limit=limit,
        offset=offset,
    )


@router.get(
    "/count",
    response_model=dict,
    summary="Count project comments",
    description="Live comments on the project, replies included.",
)
def count_comments(
    project_id: uuid.UUID,
    db: Session = Depends(get_database),
    current_user: User | None = Depends(get_optional_current_user),
) -> dict:
    project = ProjectCommentService.get_project_or_404(db, project_id)
    ProjectCommentService.require_can_view(db, project, current_user)

    return {"count": ProjectCommentService.count_comments(db, project_id)}


@router.post(
    "",
    response_model=ProjectCommentResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Post a comment or reply",
    description=(
        "Add a comment to the project discussion. Pass `parent_id` to reply. "
        "Replies are one level deep -- replying to a reply attaches to its "
        "parent instead."
    ),
)
@router.post(
    "/",
    response_model=ProjectCommentResponse,
    status_code=status.HTTP_201_CREATED,
    include_in_schema=False,
)
def create_comment(
    project_id: uuid.UUID,
    payload: ProjectCommentCreate,
    db: Session = Depends(get_database),
    current_user: User = Depends(get_current_active_user),
) -> ProjectCommentResponse:
    project = ProjectCommentService.get_project_or_404(db, project_id)
    ProjectCommentService.require_can_view(db, project, current_user)

    comment = ProjectCommentService.create_comment(db, project, payload, current_user)
    db.commit()
    db.refresh(comment)

    return _to_response(comment)


@router.patch(
    "/{comment_id}",
    response_model=ProjectCommentResponse,
    summary="Edit a comment",
    description="Authors may edit their own comments. Nobody else can.",
)
def update_comment(
    project_id: uuid.UUID,
    comment_id: uuid.UUID,
    payload: ProjectCommentUpdate,
    db: Session = Depends(get_database),
    current_user: User = Depends(get_current_active_user),
) -> ProjectCommentResponse:
    project = ProjectCommentService.get_project_or_404(db, project_id)
    ProjectCommentService.require_can_view(db, project, current_user)

    comment = ProjectCommentService.get_comment_or_404(db, project_id, comment_id)
    ProjectCommentService.require_can_edit(comment, current_user)

    comment = ProjectCommentService.update_comment(db, comment, payload)
    db.commit()
    db.refresh(comment)

    return _to_response(comment)


@router.delete(
    "/{comment_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete a comment",
    description=(
        "Soft delete. Authors may remove their own comments; a project owner "
        "may remove any comment on their project. The row is tombstoned so "
        "replies underneath keep their context."
    ),
)
def delete_comment(
    project_id: uuid.UUID,
    comment_id: uuid.UUID,
    db: Session = Depends(get_database),
    current_user: User = Depends(get_current_active_user),
) -> Response:
    project = ProjectCommentService.get_project_or_404(db, project_id)
    ProjectCommentService.require_can_view(db, project, current_user)

    comment = ProjectCommentService.get_comment_or_404(db, project_id, comment_id)
    ProjectCommentService.require_can_delete(db, project, comment, current_user)

    ProjectCommentService.delete_comment(db, comment, current_user)
    db.commit()

    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: frontend/src/api/modules/projectComments.ts
import { api } from "../client";

export interface CommentAuthor {
  id: string;
  username: string;
  first_name: string | null;
  last_name: string | null;
  profile_image: string | null;
}

export interface ProjectComment {
  id: string;
  project_id: string;
  /** Null for a top-level comment. Replies are one level deep. */
  parent_id: string | null;
  /** `"[deleted]"` when `is_deleted` is true — the real text is never sent. */
  body: string;
  /** Null on a deleted comment, so the thread does not name who wrote it. */
  author: CommentAuthor | null;
  is_edited: boolean;
  edited_at: string | null;
  is_deleted: boolean;
  created_at: string;
  updated_at: string;
}

export interface ProjectCommentThread extends ProjectComment {
  replies: ProjectComment[];
  reply_count: number;
}

export interface ProjectCommentList {
  items: ProjectCommentThread[];
  /** Top-level comments only, so it lines up with the page size. */
  total: number;
  limit: number;
  offset: number;
}

export interface ListCommentsParams {
  limit?: number;
  offset?: number;
}

export interface CreateCommentPayload {
  body: string;
  /** Omit for a top-level comment. */
  parent_id?: string | null;
}

export interface UpdateCommentPayload {
  body: string;
}

const base = (projectId: string) => `/api/v1/projects/${projectId}/comments`;

export const projectCommentsApi = {
  /** Top-level comments, newest first, each with its replies. */
  list(
    projectId: string,
    params: ListCommentsParams = {},
    signal?: AbortSignal,
  ): Promise<ProjectCommentList> {
    return api.get<ProjectCommentList>(base(projectId), {
      query: { limit: params.limit, offset: params.offset },
      signal,
    });
  },

  /** Live comments including replies — the number for a "12 comments" badge. */
  count(projectId: string, signal?: AbortSignal): Promise<{ count: number }> {
    return api.get<{ count: number }>(`${base(projectId)}/count`, { signal });
  },

  create(projectId: string, payload: CreateCommentPayload): Promise<ProjectComment> {
    // Send parent_id explicitly rather than omitting it, so the request body
    // has the same shape for a comment and a reply.
    return api.post<ProjectComment>(base(projectId), {
      body: payload.body,
      parent_id: payload.parent_id ?? null,
    });
  },

  update(
    projectId: string,
    commentId: string,
    payload: UpdateCommentPayload,
  ): Promise<ProjectComment> {
    return api.patch<ProjectComment>(`${base(projectId)}/${commentId}`, payload);
  },

  /**
   * Soft delete. The comment stays in the listing as a tombstone so replies
   * underneath it keep their context, so callers should refetch rather than
   * removing the row locally.
   */
  remove(projectId: string, commentId: string): Promise<void> {
    return api.delete<void>(`${base(projectId)}/${commentId}`);
  },
};
